### ISSUE TITLE
Interest in an unsafe Aff to Eff (with timeout)?

### ISSUE THREAD

**Issue by pm-1z** [slamdata/purescript-aff#66]
I wrote a little thing in purescript to use on AWS Lambda, and I really wanted to use Aff for the internal AWS async calls (mostly to S3) instead of Eff. However, when I tried using something like:

    f :: forall e . Foreign -> Aff (... | e) -> Response

as the Lambda handler, Aff tried to execute an `undefined` success handler and then finally killed itself by trying to execute an `undefined` error handler. As someone mentioned to me on StackOverflow, this was a side effect of me trying to pretend Aff could be called directly and return a non-Unit value.

Long story short, I wrote something that lets me supply a timeout and an `Aff e a`, and gives me an `Eff e a` :

    type TimeoutMs = Int
    f : forall e a . TimeoutMs -> Aff e a -> Eff e a

This seems really dirty, but there is a need for this sort of construct if you are going to be using asynchronous services on AWS Lambda. You usually want to return a value as the result of an invocation.

I guess my "issue" is around if there's any desire for this kind of construct to reside in purescript-aff (I'd be happy to submit a PR to get my code in, although it's rather trivial), or since I'm really new to purescript itself, if there's a better way to build this kind of thing without the unfortunate FFI'ed javascript setTimeout busy wait.

**Comment by garyb** [comment]
How can you get an `Eff e a` out of an async `Aff` without providing the `a` it will return?

**Comment by pm-1z** [comment]
Well, I use it inside of a larger function, and the `Eff` from the function definition, as well as the `Aff` it runs, has a corresponding `a`.

I have a simple example:

```
readFile :: forall e. Foreign -> Eff (err :: EXCEPTION , console :: CONSOLE | e) Response
readFile reqF = unsafeRunAff 1000 do
  case (read reqF) of
    Right req -> do
      res <- getObjectForUser req
      pure $ Response { status: 200, body: res }

    Left err -> liftEff do
      C.log $ "error " <> (show err)
      pure $ Response { status: 400, body: ("invalid input: " <> (show err)) }
```

in another file

```
foreign import getObjectForUserEff :: forall e . (String -> Eff e Unit) -> DocRequest -> Eff e Unit

getObjectForUser :: forall e . DocRequest -> Aff e String
getObjectForUser req  = makeAff (\reject resolve -> getObjectForUserEff resolve req)
```

I haven't _extensively_ tested it, but the type checker seems to prevent me from returning anything other than a `Response` record.

**Comment by garyb** [comment]
Ah, so it is for a synchronous `Aff` then.

I'm not entirely sure what it is you're trying to do here if I'm honest though :wink:. Generally when you use `Aff`, the program is entirely in `Aff`, with a `launchAff` to turn it into an `Eff` so it's suitable for use as `main`.

You can freely mix `Eff` actions into `Aff` by using `liftEff`, but there's never going to be a general purpose safe way to lower an `Aff`-with-result into an `Eff`, and I can't really think of a case where it would be strictly needed.

If you're just using `Aff` for its error-handling capabilities then there are better options for handling it (`MonadError`, `Either`, `Maybe`, or even just `catchException`) that don't come along with the implication of asynchronicity.

**Comment by pm-1z** [comment]
Yes, it's to provide guaranteed termination of an asynchronous task using a watchdog, essentially. The AWS Lambda programming model requires you supply it a handler function (which is not the main function, that actually gets called on the first invocation of any specific Lambda for setting up any internal state needed by the handler). Since Lambdas are not long running tasks like a traditional node.js server is, and Lambdas will automatically propagate the return value from the handler to other AWS destinations (S3, Kinesis, API Gateway, etc) the handler function really shouldn't return `Unit`.

The problem I was running into was if I made my Lambda handler (in this case the `readFile` function) simply return an `Aff e Response` I would get exceptions as the _makeAff function kept trying to execute `undefined` success and error functions. Since there isn't a way to get `Aff` to ever return a value, and I needed to directly return a value from an async task, I made this horrid thing.

Of course writing that out again, I just realized that since the entry point to the handler is in plain old Javascript and it calls the purescript handler, I could pass the success and error handlers in manually (something like this `PS['Main'].readFile(event)(function (v) { /* handle Aff success */ }, function (e) { /* handle Aff error */ })` and not have to build this at all. 

That's probably the cleaner route to take so unless you have any interest in this thing I'll close the issue.
